OCPBUGS-51261: controller: RTE `ConfigMap` predicate

Tal-or opened the pull request:
A continuation of #1216 which turns out to solve the problem only partially.

The ConfigMap generated by Kubelet controller
should be watched by NUMAResourceoperator controller,
so it will be able to act upon a ConfigMap change.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
controller: configmap predicate

The configmap generated by Kubelet controller
should be watched by NUMAResourceoperator controller,
so it will be able to act upon a configmap change.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
e2e: test correct TM configuration

Add an e2e test to make sure NRTs are reflecting the correct TM configuration

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or commented:
/retest

rbaturov commented on test/e2e/install/install_test.go:
```diff
@@ -356,9 +360,80 @@ var _ = Describe("[Install] durability", Serial, func() {
+					nrt := &updatedNrtObjs.Items[i]
+					// in this specific test deployment,
+					// the same configuration should apply to all NRT objects
+					matchingResult := checkTopologyManagerConfigMatching(nrt, &cfg)
```
I think a better name instead of matchingResult could be `mismatchErrors` or `misconfigurationErrors`, since the function only returns details when a mismatch is found. What do you think?
I would change it inside the `checkTopologyManagerConfigMatching` function as well.

Tal-or replied:
Thanks. Fixed

swatisehgal reviewed (commented):
Thanks for the PR and also adding e2e test.

I understand that the e2e test you have added allows us to ensure that we have expected Topology Manager configuration and I think this is of value but I am failing to understand how it is proof of the fact that we have resolved the issue at hand (NROP controller not creating new CM on node groups updates). Can you please explain?

Can we add an e2e test for validating that the code changes proposed here resolve the main issue and that new CMs are created if we point to a different node group. Adding an e2e would also automate the verification process of the bug.

Tal-or commented:
> Thanks for the PR and also adding e2e test.
> 
> I understand that the e2e test you have added allows us to ensure that we have expected Topology Manager configuration and I think this is of value but I am failing to understand how it is proof of the fact that we have resolved the issue at hand (NROP controller not creating new CM on node groups updates). Can you please explain?
> 
> Can we add an e2e test for validating that the code changes proposed here resolve the main issue and that new CMs are created if we point to a different node group. Adding an e2e would also automate the verification process of the bug.

The bug describes a symptom, but not the root cause. 
The root cause was that NROP controller is not watching for CM changes that are done by the Kubelet controller. This is  why we've added the predicate for.

With that being said it might be useful to add a test that mimic the bug description exactly. We should add it in the serial suite though since it would be too complex as an u/s test.

swatisehgal commented:
> With that being said it might be useful toadd a test that mimic the bug description exactly. We should add it in the serial suite though since it would be too complex as an u/s test.

Sure, addition to the serial suite sounds reasonable. Do you prefer to do that in the same PR or a follow-up? I am fine either way.

shajmakh commented on internal/controller/numaresourcesoperator_controller.go:
```diff
@@ -672,6 +684,36 @@ func (r *NUMAResourcesOperatorReconciler) mcpToNUMAResourceOperator(ctx context.
 	return requests
 }
 
+func (r *NUMAResourcesOperatorReconciler) configMapToNUMAResourceOperator(ctx context.Context, cmObj client.Object) []reconcile.Request {
```
why do we need this if `configMapPredicates` is already filtering the config maps based on the NROP label?

Tal-or replied:
`configMapToNUMAResourceOperator` is a function which converts the `ConfigMap` (A) events  to `NUMAResourceOperator` requests, because the `NUMAResourcesOperatorReconciler` expects to reconcile the NUMAResourceOperator CR and not a `ConfigMap` objects

The configMapPredicates is under filter layer that minimizes the `ConfigMap` events that gets into the `configMapToNUMAResourceOperator` function.

shajmakh replied:
ack I see how builder Watches is expecting an event handler and options, thanks for clarifying

Tal-or committed:
kubeletcontroller: nrop predicate

Kubelet controller tracks KubeletConfig objects on the system and creates a configmap
data for RTE from them.

There's a subtle bug when a new KubeletConfig gets created on the system,
the controller cannot create the configmap data if a new nodeGroup
was not already configured in the nrop object:
https://github.com/openshift-kni/numaresources-operator/blob/597d9df3ffac7b084fefef8192f28c414c9f3087/internal/controller/kubeletconfig_controller.go#L239

Hence, we should watch for nrop CR changes under the nodeGroup and
retrigger the recociliation loop.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or commented:
split the serial tests to https://github.com/openshift-kni/numaresources-operator/pull/1266 since we don't need to backport them

shajmakh commented on internal/controller/kubeletconfig_controller.go:
```diff
@@ -155,9 +159,22 @@ func (r *KubeletConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+			nroNew := e.ObjectNew.(*nropv1.NUMAResourcesOperator)
+			oldNodeGroups := nroOld.Spec.NodeGroups
+			newNodeGroups := nroNew.Spec.NodeGroups
+			// if nodeGroups number has changed,
+			// it means that we should iterate and create/delete ConfigMap data for RTE pods
+			return len(oldNodeGroups) != len(newNodeGroups)
```
what about the case where the NROP CR was updated with different nodegroup but maintained same amount of entries in spec.nodegroup? in such case the condition would be false but in reality the groups changed

Tal-or replied:
Right. let's do a simple compare instead

swatisehgal reviewed (commented):
LGTM from my side.

Will let @shajmakh add the label once she is happy enough.

shajmakh commented on internal/controller/kubeletconfig_controller.go:
```diff
@@ -20,12 +20,14 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"k8s.io/apimachinery/pkg/api/equality"
```
nit: misplaced import

shajmakh commented:
/lgtm
Thanks for handing this
